File: leetcode/Explore-August/week4/August24/soln.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

def rec_sumOfLeftLeaves(node, ans):
    if(node):
        if(node.left):
            if(not(node.left.left) and not(node.left.right)):
                ans[0] += node.left.val
            else:
                rec_sumOfLeftLeaves(node.left, ans)
        rec_sumOfLeftLeaves(node.right, ans)
    

class Solution:
    def sumOfLeftLeaves(self, root: TreeNode) -> int:
        ans = [0]
        rec_sumOfLeftLeaves(root, ans)
        return ans[0]
        
        

File: leetcode/Explore-August/week5/August31/soln.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

typedef struct TreeNode node_t;

node_t* inorder_succ(node_t* node)
{
    node_t* succ = NULL;
    node_t* parent = node;
    if(node -> right)
    {
        succ = node -> right;
    }
    while(succ -> left)
    {
        parent = succ;
        succ = succ -> left;
    }
    if(succ && succ -> val < parent -> val)
    {
        parent -> left = succ -> right;
    }
    else if (succ && succ -> val > parent -> val)
    {
        parent -> right = succ -> right;
    }
    return succ;
}


struct TreeNode* deleteNode(struct TreeNode* root, int key){
    if(!root)
    {
        return root;
    }
    
    if(root -> val == key)
    {
        node_t* temp = root;
        if(!root -> right)
        {
            root = root -> left;
            free(temp);
            return root;
        }
    }
    
    node_t* node = root;
    node_t* parent = NULL;
    while(node && node -> val != key)
    {
        parent = node;
        if(key < node -> val)
        {
            node = node -> left;
        }
    
        else
        {
            node = node -> right;
        }
    }
    if(node == NULL)
    {
        return root;
    }
    
    if(node -> right == NULL && parent)
    {
        if(node -> val < parent -> val)
        {
            parent -> left = node -> left;
        }
        else
        {
            parent -> right = node -> left;
        }
        return root;
    }
    node_t* succ = inorder_succ(node);
    node -> val = succ -> val;
    free(succ);
    return root;
}
